[GTK][WPE][Skia] Introduce threaded GPU painting mode

nikolaszimmermann opened the pull request:
#### 8566d855663a3dadd06c1562746e95344fb5af18
<pre>
[GTK][WPE][Skia] Introduce threaded GPU painting mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=280677">https://bugs.webkit.org/show_bug.cgi?id=280677</a>

Reviewed by Carlos Garcia Campos.

This patch introduces a threaded rendering mode also for the GPU
accelerated rendering, and activates it by default. The change aims to
free the main thread: currently accelerated rendering in the
Gtk/WPE ports blocks the main thread -- move to a DisplayList
record &amp; replay approach, to make the GPU tile painting asynchronous,
which has proven to work successfully for CPU rendering already.

In GPU rendering mode we previously didn&apos;t use any threading, and thus
painted tile after tile, without any parallelization. GPU painting does
requires a non-neglible amount of CPU work, to prepare the GPU painting
(computing vertex data, etc.). This slowed down the current GPU
rendering architecture so much, that threaded CPU rendering on embedded
devices was usually faster than GPU rendering. However CPU rendering is
not per-se faster, it&apos;s just the naive implementation we&apos;re currently
using that makes the GPU rendering appear slower than the CPU rendering.

A considerable amount of time was spent to adapt the ImageBuffer and
NativeImage rendering code to cope with threaded GPU painting: we cannot
re-use the SkSurface/SkImages produced on main thread directly on the
worker threads -- instead we have to extract the underlying
GrBackendRenderTarget/GrBackendTexture, and re-wrap them in
SkSurfaces/SkImages on the worker thread.

The SkiaThreadedPaintingPool class was renamed to SkiaPaintingEngine,
since it now takes care of both unaccelerated &amp; accelerated rendering,
either in multi-threaded, or main thread mode.

The environment variable WEBKIT_SKIA_PAINTING_THREADS was renamed to
WEBKIT_SKIA_CPU_PAINTING_THREADS, since it only controls the amount
of CPU rendering threads that are used (default: nCores/2).

Introduce WEBKIT_SKIA_GPU_PAINTING_THREADS (default: 1), to control
the amount of GPU worker threads.

The different painting modes can be selected at runtime:

1) GPU-accelerated multi-threaded rendering: (now default!)
   - WEBKIT_SKIA_GPU_PAINTING_THREADS&gt;0
     (or unset, then default 1 is used)
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING unset

2) GPU-accelerated main thread rendering: (previous default)
   - WEBKIT_SKIA_GPU_PAINTING_THREADS=0
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING unset

3) CPU multi-threaded rendering:
   - WEBKIT_SKIA_CPU_PAINTING_THREADS&gt;0
     (or unset, then default nCores/2 is used)
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING=1

4) CPU main thread rendering:
   - WEBKIT_SKIA_CPU_PAINTING_THREADS=0
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING=1

Covered by existing tests.

* LayoutTests/css3/masking/mask-base64.html:
* LayoutTests/css3/masking/mask-luminance-svg.html:
* LayoutTests/svg/filters/filter-on-root-tile-boundary.html:
* LayoutTests/svg/repaint/buffered-rendering-static-image.html:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::skiaGrContext const):
(WebCore::ImageBuffer::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::backendInfo const):
(WebCore::ImageBuffer::backendInfo): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::skiaGrContext const):
(WebCore::ImageBufferBackend::parameters const):
(WebCore::ImageBufferBackend::parameters): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::skiaGrContext const):
(WebCore::NativeImageBackend::copyAcceleratedNativeImageBorrowingBackendTexture const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawSkiaImage):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer):
(WebCore::GraphicsContextSkia::drawSkiaPattern):
(WebCore::GraphicsContextSkia::drawPattern):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompletion):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::PlatformImageNativeImageBackend::waitForAcceleratedRenderingFenceCompletion):
(WebCore::PlatformImageNativeImageBackend::skiaGrContext const):
(WebCore::PlatformImageNativeImageBackend::copyAcceleratedNativeImageBorrowingBackendTexture const):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp: Added.
(WebCore::SkiaPaintingEngine::SkiaPaintingEngine):
(WebCore::SkiaPaintingEngine::create):
(WebCore::SkiaPaintingEngine::recordDisplayList const):
(WebCore::SkiaPaintingEngine::paintIntoGraphicsContext const):
(WebCore::SkiaPaintingEngine::paintDisplayListIntoBuffer):
(WebCore::SkiaPaintingEngine::paintGraphicsLayerIntoBuffer const):
(WebCore::canPerformAcceleratedRendering):
(WebCore::SkiaPaintingEngine::renderingMode const):
(WebCore::SkiaPaintingEngine::threadedRenderingMode const):
(WebCore::SkiaPaintingEngine::createBuffer const):
(WebCore::SkiaPaintingEngine::paintLayer):
(WebCore::SkiaPaintingEngine::postPaintingTask):
(WebCore::SkiaPaintingEngine::performPaintingTask):
(WebCore::SkiaPaintingEngine::numberOfCPUPaintingThreads):
(WebCore::SkiaPaintingEngine::numberOfGPUPaintingThreads):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h: Renamed from Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h.
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp: Removed.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.cpp:
(WebCore::CoordinatedTileBuffer::beginPainting):
(WebCore::CoordinatedTileBuffer::completePainting):
(WebCore::CoordinatedTileBuffer::waitUntilPaintingComplete):
(WebCore::CoordinatedAcceleratedTileBuffer::completePainting):
(WebCore::CoordinatedAcceleratedTileBuffer::waitUntilPaintingComplete):
(WebCore::CoordinatedUnacceleratedTileBuffer::beginPainting): Deleted.
(WebCore::CoordinatedUnacceleratedTileBuffer::completePainting): Deleted.
(WebCore::CoordinatedUnacceleratedTileBuffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
(WebKit::LayerTreeHost::~LayerTreeHost):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:

Canonical link: <a href="https://commits.webkit.org/287060@main">https://commits.webkit.org/287060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78719774fc0c3215f2564b105b77a8bf9282b7cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82825 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80300 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5494 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/19135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81234 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51208 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/68065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/41531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48561 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/24786 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27766 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84189 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5532 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3739 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5689 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17135 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12697 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10944 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5481 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/8233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5470 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8902 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7259 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -39,17 +43,28 @@ namespace WebCore {
+    if (ProcessCapabilities::canUseAcceleratedBuffers()) {
+        m_texturePool = makeUnique<BitmapTexturePool>();
+
+        if (numberOfGPUThreads > 0)
```
```suggestion
        if (numberOfGPUThreads)
```

nikolaszimmermann replied:
Well it can be null, in that case we want to avoid creating the worker pool.

nikolaszimmermann replied:
Heh, forget about it. It's unsigned, so this is equivalent.

nikolaszimmermann replied:
Fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -39,17 +43,28 @@ namespace WebCore {
+
+        if (numberOfGPUThreads > 0)
+            m_gpuWorkerPool = WorkerPool::create("SkiaGPUWorker"_s, numberOfGPUThreads);
+    } else if (numberOfCPUThreads > 0)
```
```suggestion
    } else if (numberOfCPUThreads)
```

nikolaszimmermann replied:
Fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -60,30 +75,92 @@ std::unique_ptr<DisplayList::DisplayList> SkiaThreadedPaintingPool::recordDispla
+bool SkiaThreadedPaintingPool::canPerformAcceleratedRendering() const
+{
+    return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
+}
```
This has the side effect of making skia context current, I think it would be better to check if there's a skia gl context created.

nikolaszimmermann replied:
Sounds reasonable, will adapt.

nikolaszimmermann replied:
Oh I remember now: I want to fallback to CPU rendering, if makeContextCurrent() actually fails.
Since that function is used to decide whether to use GPU rendering or not, kaing the context current has to be part of this function -- it's on purpose.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -60,30 +75,92 @@ std::unique_ptr<DisplayList::DisplayList> SkiaThreadedPaintingPool::recordDispla
-            static thread_local RefPtr<ControlFactory> s_controlFactory;
-            if (!s_controlFactory)
-                s_controlFactory = ControlFactory::create();
+    auto* workerPool = renderingMode == RenderingMode::Accelerated ? m_gpuWorkerPool.get() : m_cpuWorkerPool.get();
```
Better use a RefPtr here instead of  a raw pointer

nikolaszimmermann replied:
Why unnecessary ref'ing? We're only using it the line after.

carlosgcampos replied:
Just to void raw pointers, you can also move the assert before and then use a reference.

nikolaszimmermann replied:
Fair enough, let's do that.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -60,30 +75,92 @@ std::unique_ptr<DisplayList::DisplayList> SkiaThreadedPaintingPool::recordDispla
+            if (buffer->isBackedByOpenGL())
+                WTFBeginSignpost(canvas, PaintTile, "Skia/GPU threaded, dirty region %ix%i+%i+%i", dirtyRect.x(), dirtyRect.y(), dirtyRect.width(), dirtyRect.height());
+            else
+                WTFBeginSignpost(canvas, PaintTile, "Skia/CPU threaded, dirty region %ix%i+%i+%i", dirtyRect.x(), dirtyRect.y(), dirtyRect.width(), dirtyRect.height());
```
```suggestion
            WTFBeginSignpost(canvas, PaintTile, "Skia/%s threaded, dirty region %ix%i+%i+%i", buffer->isBackedByOpenGL() ? "GPU" : CPU", dirtyRect.x(), dirtyRect.y(), dirtyRect.width(), dirtyRect.height());
```

nikolaszimmermann replied:
Will adapt, thanks.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h:
```diff
@@ -45,18 +48,31 @@ class SkiaThreadedPaintingPool {
     static std::unique_ptr<SkiaThreadedPaintingPool> create();
 
-    void postPaintingTask(Ref<Nicosia::Buffer>&, const TiledBackingStore&, const CoordinatedGraphicsLayer&, const IntRect& dirtyRect);
+    static unsigned numberOfCPUPaintingThreads();
+    static unsigned numberOfGPUPaintingThreads();
```
Do these need to be public?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
```diff
@@ -648,6 +653,12 @@ void WebKitProtocolHandler::handleGPU(WebKitURISchemeRequest* request)
+#if USE(SKIA)
+            addTableRow(hardwareAccelerationObject, "Default painting mode"_s, ProcessCapabilities::canUseAcceleratedBuffers() ? "GPU"_s : "CPU"_s);
+            addTableRow(hardwareAccelerationObject, "CPU painting threads"_s, String::number(SkiaThreadedPaintingPool::numberOfCPUPaintingThreads()));
+            addTableRow(hardwareAccelerationObject, "GPU painting threads"_s, String::number(SkiaThreadedPaintingPool::numberOfGPUPaintingThreads()));
```
Now I see why they are public.

carlosgcampos reviewed (commented):
The approach looks good to me

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -427,6 +427,24 @@ void ImageBuffer::waitForAcceleratedRenderingFenceCompletion()
         backend->waitForAcceleratedRenderingFenceCompletion();
 }
+
+const GrDirectContext* ImageBuffer::directContext() const
```
I would call this skiaGrContext for consistency with PlatformDisplay

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBufferBackend.h:
```diff
@@ -136,6 +140,9 @@ class ImageBufferBackend {
     virtual void finishAcceleratedRenderingAndCreateFence() { }
     virtual void waitForAcceleratedRenderingFenceCompletion() { }
+
+    virtual const GrDirectContext* directContext() const { return nullptr; }
```
Same here skiaGrContext

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -427,6 +427,24 @@ void ImageBuffer::waitForAcceleratedRenderingFenceCompletion()
+    ASSERT(renderingMode() == RenderingMode::Accelerated);
+
+    auto* backend = ensureBackend();
+    ASSERT(backend);
```
So, in directContext we handle backend being null, but here we assert instead. What's the difference?

nikolaszimmermann replied:
Currently this always holds true, but it doesn't have to be -- I will change it.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -245,9 +250,37 @@ static SkSamplingOptions toSkSamplingOptions(InterpolationQuality quality)
 {
     auto image = nativeImage.platformImage();
+
+    auto* directContext = SkImages::GetContext(image);
```
```suggestion
    auto* imageGrContext = SkImages::GetContext(image);
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -245,9 +250,37 @@ static SkSamplingOptions toSkSamplingOptions(InterpolationQuality quality)
+static sk_sp<SkImage> skiaImageFromNativeImage(GraphicsContextSkia& context, NativeImage& nativeImage)
 {
     auto image = nativeImage.platformImage();
+
```
We should null check image here, and return early if it's nullptr.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -245,9 +250,37 @@ static SkSamplingOptions toSkSamplingOptions(InterpolationQuality quality)
 }
 
-void GraphicsContextSkia::drawNativeImageInternal(NativeImage& nativeImage, const FloatRect& destRect, const FloatRect& srcRect, ImagePaintingOptions options)
+static sk_sp<SkImage> skiaImageFromNativeImage(GraphicsContextSkia& context, NativeImage& nativeImage)
```
Maybe this could be a method of NativeImage? NativeImage::crossThreadPlatformImage or something like that?

nikolaszimmermann replied:
Maybe not NativeImage, but NativeImageBackend -- I'll come up with a symmetric API for ImageBuffer/NativeImage.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -303,6 +336,23 @@ void GraphicsContextSkia::drawNativeImageInternal(NativeImage& nativeImage, cons
 
+void GraphicsContextSkia::drawFilteredImageBuffer(ImageBuffer* sourceImage, const FloatRect& sourceImageRect, Filter& filter, FilterResults& results)
+{
+    if (sourceImage && sourceImage->directContext() && sourceImage->directContext() != PlatformDisplay::sharedDisplay().skiaGrContext()) {
```
directContext is not a simple getter, maybe it would be better to call it once?

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -953,7 +1003,7 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
         return;
 
-    auto image = nativeImage.platformImage();
+    auto image = skiaImageFromNativeImage(*this, nativeImage);
```
I wonder if we always want platformImage to return a valid image on any thread instead.

nikolaszimmermann replied:
That will be empty when painting it though...

carlosgcampos replied:
What do you mean? What will be empty?

nikolaszimmermann replied:
If you attempt to use the image without re-wrapping the SkSurface the painted content is empty (nothing visible)

carlosgcampos replied:
That's why I wondered if platformImage should always return a valid image, so the caller doesn't need to think if it needs to use skiaImageFromNativeImage, because platformImage also returns an skia image from the NativeImage

nikolaszimmermann replied:
Hmm good question. If native image would directly execute the wrapping logic, it would be very inefficient to call platformImage() multiple times in a row.

I will check all call sites and check if this is feasible...

carlosgcampos replied:
If that's not the case I would still move the code to NativeImage with a better name, because right now NativeImage::platformImage and skiaImageFromNativeImage are the same if you just look at the name, it's not obvios what to use. That's why I suggested NativeImage::crossThreadPlatformImage

nikolaszimmermann replied:
I've introduced `NativeImageBackend::copyAcceleratedNativeImageBorrowingBackendTexture()`, just like we already had `ImageBufferBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget()`.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -135,6 +147,26 @@ void ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompleti
     m_fence = nullptr;
 }
 
+RefPtr<ImageBuffer> ImageBufferSkiaAcceleratedBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget(const ImageBuffer* imageBuffer) const
```
ImageBuffer is used without null check, this should receive a reference instead, or a Ref<>

nikolaszimmermann replied:
Thanks for catching this, `const ImageBuffer&` is way more suitable. Ref<> is unncessary, to avoid ref-count churn.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -135,6 +147,26 @@ void ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompleti
+    if (!surface || !surface->getCanvas())
+        return nullptr;
+
+    auto backend = ImageBufferSkiaAcceleratedBackend::create(parameters(), { }, WTFMove(surface));
```
So, this new create() is only used here, I guess we could just call new here, no?

nikolaszimmermann replied:
The original create() function uses the new constructor as well:

```
std::unique_ptr<ImageBufferSkiaAcceleratedBackend> ImageBufferSkiaAcceleratedBackend::create(const Parameters& parameters, const ImageBufferCreationContext& creationContext)
{
....
    auto surface = SkSurfaces::RenderTarget(grContext, skgpu::Budgeted::kNo, imageInfo, PlatformDisplay::sharedDisplay().msaaSampleCount(), kTopLeft_GrSurfaceOrigin, &properties);
....
    return create(parameters, creationContext, WTFMove(surface));
}
```

Therefore I think it's better to keep the `create()` variant taking a `SkSurface&&`.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
```diff
@@ -57,12 +57,16 @@ class ImageBufferSkiaAcceleratedBackend final : public ImageBufferSkiaSurfaceBac
     RefPtr<GraphicsLayerContentsDisplayDelegate> m_layerContentsDisplayDelegate;
 #endif
 
+    const GrDirectContext* m_directContext { nullptr };
```
I would call this m_grContext

nikolaszimmermann replied:
It's called `m_skiaGrContext` in `PlatformDisplaySkia`, I chose the same name.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -28,90 +28,27 @@
-    paintBuffer(buffer.get());
-    WTFEndSignpost(this, PaintTile);
-    return buffer;
+    return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
```
So, in the end we are always calling the same function but with different rendering mode, so I would add helper function or lambda, and just call `paintingEngine.performPaintingTask(renderingMode(), *this, dirtyRect);`

nikolaszimmermann replied:
Hmm, what should `renderingMode()` be? A `CoordinatedGraphicsLayer` has no rendering mode per-se.

Maybe it's a bit hard to see in the diff, but we are first determining the rendering mode, by checking conditions in `SkiaPaintingEngine`, before we call `performPaintingTask`:

```
Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
{
    auto& paintingEngine = m_coordinator->skiaPaintingEngine();

    // ### Asynchronous rendering on worker threads ###
    if (paintingEngine.canPerformThreadedAcceleratedRendering())
        return paintingEngine.postPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);

    if (paintingEngine.canPerformThreadedUnacceleratedRendering())
        return paintingEngine.postPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);

    // ### Synchronous rendering on main thread ###
    if (paintingEngine.canPerformAcceleratedRendering())
        return paintingEngine.performPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);

    return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
}
```

And we also do it in a specific order. First mt-GPU, then mt-CPU, then GPU, then CPU.

But  you are right, that we could just call: `paintingEngine.performPaintingTask(*this, dirtyRect)` (without a rendering mode specified, and let `SkiaPaintingEngine` figure it out, based on the same logic).

nikolaszimmermann replied:
I reworked the logic, instead of the three bool accessors to determine the painting mode (threaded/main thread, CPU/GPU) I came up with this:

```
static bool canPerformAcceleratedRendering()
{
    return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext();
}

RenderingMode SkiaPaintingEngine::renderingMode() const
{
    if (canPerformAcceleratedRendering())
        return RenderingMode::Accelerated;

    return RenderingMode::Unaccelerated;
}

std::optional<RenderingMode> SkiaPaintingEngine::threadedRenderingMode() const
{
    if (m_gpuWorkerPool && canPerformAcceleratedRendering())
        return RenderingMode::Accelerated;

    if (m_cpuWorkerPool)
        return RenderingMode::Unaccelerated;

    return std::nullopt;
}
```

and a new `paintLayer` method:

```
Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect)
{
    // ### Asynchronous rendering on worker threads ###
    if (auto renderingMode = SkiaPaintingEngine::threadedRenderingMode())
        return postPaintingTask(renderingMode.value(), layer, dirtyRect);

    // ### Synchronous rendering on main thread ###
    return performPaintingTask(renderingMode(), layer, dirtyRect);
}
```

This makes `paintTile` trivial:

```
Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
{
    auto& paintingEngine = m_coordinator->skiaPaintingEngine();
    return paintingEngine.paintLayer(*this, dirtyRect);
}
```

carlosgcampos replied:
> Hmm, what should `renderingMode()` be? A `CoordinatedGraphicsLayer` has no rendering mode per-se.

A simple static helper or a lambda. 

> Maybe it's a bit hard to see in the diff, but we are first determining the rendering mode, by checking conditions in `SkiaPaintingEngine`, before we call `performPaintingTask`:
> 
> ```
> Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
> {
>     auto& paintingEngine = m_coordinator->skiaPaintingEngine();
> 
>     // ### Asynchronous rendering on worker threads ###
>     if (paintingEngine.canPerformThreadedAcceleratedRendering())
>         return paintingEngine.postPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
> 
>     if (paintingEngine.canPerformThreadedUnacceleratedRendering())
>         return paintingEngine.postPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
> 
>     // ### Synchronous rendering on main thread ###
>     if (paintingEngine.canPerformAcceleratedRendering())
>         return paintingEngine.performPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
> 
>     return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
> }
> ```

So, the helper would be exactly the same, but instead of calling the paint method, it would return the rendering mode. What I didn't notice in the diff is that we are not actually calling the same paint method, but post/perform, so it's not really possible, sorry for the noise.

> And we also do it in a specific order. First mt-GPU, then mt-CPU, then GPU, then CPU.
> 
> But you are right, that we could just call: `paintingEngine.performPaintingTask(*this, dirtyRect)` (without a rendering mode specified, and let `SkiaPaintingEngine` figure it out, based on the same logic).

That would make paintTile simpler indeed.

carlosgcampos reviewed (approved):
Looks great! I have just a few comments/questions.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
```diff
@@ -41,6 +41,7 @@ class ImageBufferSkiaAcceleratedBackend final : public ImageBufferSkiaSurfaceBac
     WTF_MAKE_NONCOPYABLE(ImageBufferSkiaAcceleratedBackend);
 public:
     static std::unique_ptr<ImageBufferSkiaAcceleratedBackend> create(const Parameters&, const ImageBufferCreationContext&);
+    static std::unique_ptr<ImageBufferSkiaAcceleratedBackend> create(const Parameters&, const ImageBufferCreationContext&, sk_sp<SkSurface>&&);
```
If you keep this create, I guess it could be private.

nikolaszimmermann replied:
Good point, will fix and then land! Thanks @carlosgcampos.

nikolaszimmermann committed:
[GTK][WPE][Skia] Introduce threaded GPU painting mode
https://bugs.webkit.org/show_bug.cgi?id=280677

Reviewed by Carlos Garcia Campos.

This patch introduces a threaded rendering mode also for the GPU
accelerated rendering, and activates it by default. The change aims to
free the main thread: currently accelerated rendering in the
Gtk/WPE ports blocks the main thread -- move to a DisplayList
record & replay approach, to make the GPU tile painting asynchronous,
which has proven to work successfully for CPU rendering already.

In GPU rendering mode we previously didn't use any threading, and thus
painted tile after tile, without any parallelization. GPU painting does
requires a non-neglible amount of CPU work, to prepare the GPU painting
(computing vertex data, etc.). This slowed down the current GPU
rendering architecture so much, that threaded CPU rendering on embedded
devices was usually faster than GPU rendering. However CPU rendering is
not per-se faster, it's just the naive implementation we're currently
using that makes the GPU rendering appear slower than the CPU rendering.

A considerable amount of time was spent to adapt the ImageBuffer and
NativeImage rendering code to cope with threaded GPU painting: we cannot
re-use the SkSurface/SkImages produced on main thread directly on the
worker threads -- instead we have to extract the underlying
GrBackendRenderTarget/GrBackendTexture, and re-wrap them in
SkSurfaces/SkImages on the worker thread.

The SkiaThreadedPaintingPool class was renamed to SkiaPaintingEngine,
since it now takes care of both unaccelerated & accelerated rendering,
either in multi-threaded, or main thread mode.

The environment variable WEBKIT_SKIA_PAINTING_THREADS was renamed to
WEBKIT_SKIA_CPU_PAINTING_THREADS, since it only controls the amount
of CPU rendering threads that are used (default: nCores/2).

Introduce WEBKIT_SKIA_GPU_PAINTING_THREADS (default: 1), to control
the amount of GPU worker threads.

The different painting modes can be selected at runtime:

1) GPU-accelerated multi-threaded rendering: (now default!)
   - WEBKIT_SKIA_GPU_PAINTING_THREADS>0
     (or unset, then default 1 is used)
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING unset

2) GPU-accelerated main thread rendering: (previous default)
   - WEBKIT_SKIA_GPU_PAINTING_THREADS=0
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING unset

3) CPU multi-threaded rendering:
   - WEBKIT_SKIA_CPU_PAINTING_THREADS>0
     (or unset, then default nCores/2 is used)
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING=1

4) CPU main thread rendering:
   - WEBKIT_SKIA_CPU_PAINTING_THREADS=0
   - WEBKIT_SKIA_ENABLE_CPU_RENDERING=1

Covered by existing tests.

* LayoutTests/css3/masking/mask-base64.html:
* LayoutTests/css3/masking/mask-luminance-svg.html:
* LayoutTests/svg/filters/filter-on-root-tile-boundary.html:
* LayoutTests/svg/repaint/buffered-rendering-static-image.html:
* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::skiaGrContext const):
(WebCore::ImageBuffer::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::backendInfo const):
(WebCore::ImageBuffer::backendInfo): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::skiaGrContext const):
(WebCore::ImageBufferBackend::parameters const):
(WebCore::ImageBufferBackend::parameters): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::skiaGrContext const):
(WebCore::NativeImageBackend::copyAcceleratedNativeImageBorrowingBackendTexture const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawSkiaImage):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawFilteredImageBuffer):
(WebCore::GraphicsContextSkia::drawSkiaPattern):
(WebCore::GraphicsContextSkia::drawPattern):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompletion):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyAcceleratedImageBufferBorrowingBackendRenderTarget const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::PlatformImageNativeImageBackend::waitForAcceleratedRenderingFenceCompletion):
(WebCore::PlatformImageNativeImageBackend::skiaGrContext const):
(WebCore::PlatformImageNativeImageBackend::copyAcceleratedNativeImageBorrowingBackendTexture const):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp: Added.
(WebCore::SkiaPaintingEngine::SkiaPaintingEngine):
(WebCore::SkiaPaintingEngine::create):
(WebCore::SkiaPaintingEngine::recordDisplayList const):
(WebCore::SkiaPaintingEngine::paintIntoGraphicsContext const):
(WebCore::SkiaPaintingEngine::paintDisplayListIntoBuffer):
(WebCore::SkiaPaintingEngine::paintGraphicsLayerIntoBuffer const):
(WebCore::canPerformAcceleratedRendering):
(WebCore::SkiaPaintingEngine::renderingMode const):
(WebCore::SkiaPaintingEngine::threadedRenderingMode const):
(WebCore::SkiaPaintingEngine::createBuffer const):
(WebCore::SkiaPaintingEngine::paintLayer):
(WebCore::SkiaPaintingEngine::postPaintingTask):
(WebCore::SkiaPaintingEngine::performPaintingTask):
(WebCore::SkiaPaintingEngine::numberOfCPUPaintingThreads):
(WebCore::SkiaPaintingEngine::numberOfGPUPaintingThreads):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h: Renamed from Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h.
* Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp: Removed.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintIntoGraphicsContext const): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.cpp:
(WebCore::CoordinatedTileBuffer::beginPainting):
(WebCore::CoordinatedTileBuffer::completePainting):
(WebCore::CoordinatedTileBuffer::waitUntilPaintingComplete):
(WebCore::CoordinatedAcceleratedTileBuffer::completePainting):
(WebCore::CoordinatedAcceleratedTileBuffer::waitUntilPaintingComplete):
(WebCore::CoordinatedUnacceleratedTileBuffer::beginPainting): Deleted.
(WebCore::CoordinatedUnacceleratedTileBuffer::completePainting): Deleted.
(WebCore::CoordinatedUnacceleratedTileBuffer::waitUntilPaintingComplete): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedTileBuffer.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
(WebKit::LayerTreeHost::~LayerTreeHost):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:

Canonical link: https://commits.webkit.org/287060@main
